clients/src/main/java/org/apache/kafka/common/requests/DeleteGroupsResponse.java

```diff
@@ -16,121 +16,72 @@
-        this.errors = errors;
+    public DeleteGroupsResponse(DeleteGroupsResponseData data) {
+        this.data = data;
     }
 
     public DeleteGroupsResponse(Struct struct) {
-        this.throttleTimeMs = struct.getOrElse(THROTTLE_TIME_MS, DEFAULT_THROTTLE_TIME);
-        Object[] groupErrorCodesStructs = struct.getArray(GROUP_ERROR_CODES_KEY_NAME);
-        Map<String, Errors> errors = new HashMap<>();
-        for (Object groupErrorCodeStructObj : groupErrorCodesStructs) {
-            Struct groupErrorCodeStruct = (Struct) groupErrorCodeStructObj;
-            String group = groupErrorCodeStruct.get(GROUP_ID);
-            Errors error = Errors.forCode(groupErrorCodeStruct.get(ERROR_CODE));
-            errors.put(group, error);
-        }
-
-        this.errors = errors;
+        short latestVersion = (short) (DeleteGroupsResponseData.SCHEMAS.length - 1);
+        this.data = new DeleteGroupsResponseData(struct, latestVersion);
     }
 
-    @Override
-    protected Struct toStruct(short version) {
-        Struct struct = new Struct(ApiKeys.DELETE_GROUPS.responseSchema(version));
-        struct.setIfExists(THROTTLE_TIME_MS, throttleTimeMs);
-        List<Struct> groupErrorCodeStructs = new ArrayList<>(errors.size());
-        for (Map.Entry<String, Errors> groupError : errors.entrySet()) {
-            Struct groupErrorCodeStruct = struct.instance(GROUP_ERROR_CODES_KEY_NAME);
-            groupErrorCodeStruct.set(GROUP_ID, groupError.getKey());
-            groupErrorCodeStruct.set(ERROR_CODE, groupError.getValue().code());
-            groupErrorCodeStructs.add(groupErrorCodeStruct);
-        }
-        struct.set(GROUP_ERROR_CODES_KEY_NAME, groupErrorCodeStructs.toArray());
-        return struct;
+    public DeleteGroupsResponse(Struct struct, short version) {
+        this.data = new DeleteGroupsResponseData(struct, version);
     }
 
     @Override
-    public int throttleTimeMs() {
-        return throttleTimeMs;
+    protected Struct toStruct(short version) {
+        return data.toStruct(version);
     }
 
     public Map<String, Errors> errors() {
-        return errors;
+        Map<String, Errors> errorMap = new HashMap<>();
+        for (DeletableGroupResult result : data.results()) {
+            errorMap.put(result.groupId(), Errors.forCode(result.errorCode()));
+        }
+        return errorMap;
     }
 
     public boolean hasError(String group) {
-        return errors.containsKey(group) && errors.get(group) != Errors.NONE;
+        DeletableGroupResult result = data.results().find(group);
+        return result != null && result.errorCode() != Errors.NONE.code();
     }
 
     public Errors get(String group) {
-        return errors.get(group);
+        return Errors.forCode(data.results().find(group).errorCode());
     }
 
     @Override
     public Map<Errors, Integer> errorCounts() {
-        return errorCounts(errors);
+        HashMap<Errors, Integer> counts = new HashMap<>();
```

Comment:
Let's use `Map` on the left side instead of `HashMap`

Reply:
sounds good